### ISSUE TITLE
Revise Observable.mapParallelOrdered

### ISSUE THREAD

**Issue by Avasil** [monix/monix#927]
[Implementation](https://github.com/monix/monix/blob/master/monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/MapParallelOrderedObservable.scala) is quite complicated and perhaps could be improved.

I would like to see benchmark vs `Observable.mapParallelUnordered` and then if replacing `java.ConcurrentQueue` with the one from `jctools` has any performance benefit.

Ideally we could also simplify implementation but I don't have anything on my mind yet

**Comment by Avasil** [comment]
In both `mapParallelUnordered` and `mapParallelOrdered` we do `f(elem).redeem(...).doOnCancel(...)` and return `Task[Unit]`. 

I wonder if there is a performance difference if we just replace it with `guaranteeCase`.

**Comment by TapanVaishnav** [comment]
@Avasil Will have a look at this one, thank you for the mention and sharing about the issue. You are the best

**Comment by TapanVaishnav** [comment]
Now that we have created the benchmark, I am starting working on different queues.
Right now, I am working with `SpmcArrayQueue`.

**Comment by Avasil** [comment]
@TapanVaishnav I would suggest `LowLevelConcurrentQueue` because available queue can be different depending on the platform. See `LowLevelConcurrentQueueBuilders`

**Comment by TapanVaishnav** [comment]
> I would suggest LowLevelConcurrentQueue because available queue can be different depending on the platform. See LowLevelConcurrentQueueBuilders

Yes, I checked this one but because this is a trait, I thought I'll start with someone already implemented(to get a hold of the problem)
> See LowLevelConcurrentQueueBuilders

Okay, I'll have a look again on `LowLevelConcurrentQueue` family.
Thanks for the suggestion :)

**Comment by Avasil** [comment]
It is used internally in `AsyncQueue` and `ConcurrentQueue` if you are looking for example
